tests/http_server/test_web_server.py

```diff
@@ -0,0 +1,154 @@
+        # if server_cert_and_key_path is not None:
+        #     self._is_https_enabled = True
+        #     self.socket = ssl.wrap_socket(self.socket,
+        #                                   certfile=server_cert_and_key_path,
+        #                                   server_side=True)
+        # else:
+        self._is_https_enabled = False
+
+    def handle_timeout(self):
+        """Overridden from SocketServer."""
+        raise RuntimeError('Timed out waiting for http request')
+
+    def get_url(self, host=None):
+        """Returns the base URL of the server."""
+        postfix = '://{}:{}'.format(host or '127.0.0.1', self.server_port)
+        if self._is_https_enabled:
+            return 'https' + postfix
+        return 'http' + postfix
+
+
+class _ThreadingServer(socketserver.ThreadingMixIn, _BaseServer):
+    """_BaseServer enhanced to handle multiple requests simultaneously"""
+    pass
+
+
+class StaticWebServer:
+    """An HTTP or HTTPS server that serves on its own thread.
+    """
+    def __init__(self):
+        """Starts the server"""
+        self._server = _ThreadingServer(self._on_request)
+        self._thread = threading.Thread(target=self._server.serve_forever)
+        self._thread.daemon = True
+        self._thread.start()
+        self._path_data_map = {}
+        self._path_maps_lock = threading.Lock()
+        self._counter = itertools.count(1)
+        self.requests = {}
+
+    def _on_request(self, handler):
+        path = handler.path.split('?')[0]
+
+        self._path_maps_lock.acquire()
+        try:
+            print("path", path)
+            if path in self._path_data_map:
+                data = self._path_data_map[path]
+                print("data", data)
+                content = data["content"]
+                handler.send_response(data["code"])
+
+                if "headers" in data:
+                    headers = data["headers"]
+                else:
+                    headers = {}
+
+                print("headers", headers)
+                print(headers.items())
+                for field, value in headers.items():
+                    print("qwe", field, value)
+                    handler.send_header(field, value)
+
+                if content is not None:
+                    handler.send_header('Content-Length', len(content))
+                handler.end_headers()
+
+                if content is not None:
+                    handler.wfile.write(content)
+                return
+            else:
+                handler.send_error(404)
+                return
+        finally:
+            self._path_maps_lock.release()
+
+    def _get_url(self, host=None):
+        """Returns the base URL of the server."""
+        return self._server.get_url(host)
+
+    def shutdown(self):
+        """Shuts down the server synchronously."""
+        self._server.shutdown()
+        self._thread.join()
+
+    def url_200(self, content="<html><body>some page</body></html>"):
```

Comment:
how would we do cross-origin URLs? I suggest we follow the helpers from WPT and have the same ones?

Reply:
We can simply extend the method with `host`, and provide specific response for specific host

Reply:
what I mean is that we can have the same helper as in WPT: https://github.com/web-platform-tests/wpt/blob/master/webdriver/tests/support/fixtures.py#L240

we probably don't need `url_200` but `def url(status=200)` and `url(status=301)` instead of `url_301`.

Reply:
It will be `url(301, {"Location": redirect_url})`. And `url_301` is a wrapper doing exactly that.

Reply:
Moved `url_${code}` our from web server to fixtures.